lib/oli/grading.ex

```diff
@@ -0,0 +1,124 @@
+  alias Oli.Delivery.Sections.SectionRoles
+
+  @doc """
+  Exports the gradebook for the provided section in CVS format
+
+  Returns a Stream which can be written to a file or other IO
+  """
+  def export_csv(%Section{} = section) do
+    {gradebook, column_labels} = generate_gradebook_for_section(section)
+
+    table_data = gradebook
+      |> Enum.map(fn %GradebookRow{user: user, scores: scores} ->
+        [
+          "#{user.first_name} #{user.last_name} (#{user.email})"
+          | Enum.map(scores, fn gradebook_score ->
+            case gradebook_score do
+              nil -> nil
+              %GradebookScore{score: score} ->
+                score
+            end
+          end)
+        ]
+      end)
+
+    # unfortunately we must go through every score to ensure out_of has been found for a column
+    # TODO: optimize this logic to bail out once an out_of has been discovered for every column
+    points_possible = gradebook
+      |> Enum.reduce([], fn %GradebookRow{scores: scores}, acc ->
+        scores
+        |> Enum.with_index
+        |> Enum.map(fn {gradebook_score, i} ->
+          case gradebook_score do
+            nil ->
+              # use existing value for column
+              Enum.at(acc, i)
+            %GradebookScore{out_of: nil} ->
+              # use existing value for column
+              Enum.at(acc, i)
+            %GradebookScore{out_of: out_of} ->
+              # replace value of existing column
+              out_of
+          end
+        end)
+      end)
+
+    points_possible = [["    Points Possible" | points_possible]]
+    column_labels = [["Student" | column_labels]]
+
+    column_labels ++ points_possible ++ table_data
+    |> CSV.encode
+  end
+
+  @doc """
+  Returns a tuple containing a list of GradebookRow for every enrolled user
+  and an ordered list of column labels
+
+  `{[%GradebookRow{user: %User{}, scores: [%GradebookScore{}, ...]}, ...], ["Quiz 1", "Quiz 2"]}`
+  """
+  def generate_gradebook_for_section(%Section{} = section) do
+    # get publication for the section
+    publication = Sections.get_section_publication!(section.id)
+
+    # get publication page resources, filtered by graded: true
+    graded_pages = Publishing.get_resource_revisions_for_publication(publication)
+      |> Map.values
+      |> Enum.filter(fn {_resource, revision} -> revision.graded == true end)
+      |> Enum.map(fn {_resource, revision} -> revision end)
+
+    # get students enrolled in the section, filter by role: student
+    students = Sections.list_enrollments(section.context_id)
+      |> Enum.filter(fn e -> e.section_role_id == SectionRoles.get_by_type("student").id end)
+      |> Enum.map(fn e -> e.user end)
+
+    # create a map of all resource accesses, keyed off resource id
+    resource_accesses = Enum.reduce(graded_pages, %{}, fn revision, acc ->
+      user_resource_access_map = Attempts.get_resource_access_for_context(revision.resource_id, section.context_id)
```

Comment:
This is a potential performance issue. Here we are doing a query for resource access records for all users in a context for a given resource - INSIDE of a loop over all graded resources in a course.  For LearnLab this list of graded pages in a course will likely be of length one or two - but eventually large courses could have dozens of graded assessments.  We may want to rework this so that we do all of this in one query (a query for all published resource ids joined with access records).

Reply:
consolidated to a single database call with multiple joins